invoices/sql_migration_test.go

```diff
@@ -0,0 +1,421 @@
+	// Set MPP total amount if MPP feature is enabled in the invoice.
+	if invoice.Terms.Features.HasFeature(lnwire.MPPRequired) {
+		htlc.MppTotalAmt = invoice.Terms.Value
+	}
+
+	// Set the HTLC state and resolve time based on the invoice state.
+	switch invoice.State {
+	case ContractSettled:
+		htlc.State = HtlcStateSettled
+		htlc.ResolveTime = randInvoiceTime(invoice)
+
+	case ContractCanceled:
+		htlc.State = HtlcStateCanceled
+		htlc.ResolveTime = randInvoiceTime(invoice)
+
+	case ContractAccepted:
+		htlc.State = HtlcStateAccepted
+	}
+
+	// Add randomized custom records to the HTLC.
+	htlc.CustomRecords = make(record.CustomSet)
+	numRecords := rapid.IntRange(0, 5).Draw(t, "numRecords")
+	for i := 0; i < numRecords; i++ {
+		key := rapid.Uint64Range(
+			record.CustomTypeStart, 1000+record.CustomTypeStart,
+		).Draw(t, "customRecordKey")
+		value := []byte(randomStringRapid(t, 10))
+		htlc.CustomRecords[key] = value
+	}
+
+	// Generate a unique HTLC ID and assign it to a channel ID.
+	htlcID := atomic.AddUint64(&testHtlcIDSequence, 1)
+	randChanID := lnwire.NewShortChanIDFromInt(htlcID % 5)
+
+	circuitKey := models.CircuitKey{
+		ChanID: randChanID,
+		HtlcID: htlcID,
+	}
+
+	return circuitKey, htlc
+}
+
+// generateInvoiceHTLCsRapid generates all HTLCs for an invoice, including AMP
+// HTLCs if applicable, using rapid for randomization of HTLC count and
+// distribution.
+func generateInvoiceHTLCsRapid(t *rapid.T, invoice *Invoice) {
+	mpp := invoice.Terms.Features.HasFeature(lnwire.MPPRequired)
+
+	// Use rapid to determine the number of HTLCs based on invoice state and
+	// MPP feature.
+	numHTLCs := 1
+	if invoice.State == ContractOpen {
+		numHTLCs = 0
+	} else if mpp {
+		numHTLCs = rapid.IntRange(1, 10).Draw(t, "numHTLCs")
+	}
+
+	total := invoice.Terms.Value
+
+	// Distribute the total amount across the HTLCs, adding any remainder to
+	// the last HTLC.
+	if numHTLCs > 0 {
+		amt := total / lnwire.MilliSatoshi(numHTLCs)
+		remainder := total - amt*lnwire.MilliSatoshi(numHTLCs)
+
+		for i := 0; i < numHTLCs; i++ {
+			if i == numHTLCs-1 {
+				// Add remainder to the last HTLC.
+				amt += remainder
+			}
+
+			// Generate an HTLC with a random circuit key and add it
+			// to the invoice.
+			circuitKey, htlc := randHTLCRapid(t, invoice, amt)
+			invoice.Htlcs[circuitKey] = htlc
+		}
+	}
+}
+
+// generateAMPHtlcsRapid generates AMP HTLCs for an invoice using rapid to
+// randomize various parameters of the HTLCs in the AMP set.
+func generateAMPHtlcsRapid(t *rapid.T, invoice *Invoice) {
+	// Randomly determine the number of AMP sets (1 to 5).
+	numSetIDs := rapid.IntRange(1, 5).Draw(t, "numSetIDs")
+	settledIdx := uint64(1)
+
+	for i := 0; i < numSetIDs; i++ {
+		var setID SetID
+		_, err := crand.Read(setID[:])
+		require.NoError(t, err)
+
+		// Determine the number of HTLCs in this set (1 to 5).
+		numHTLCs := rapid.IntRange(1, 5).Draw(t, "numHTLCs")
+		total := invoice.Terms.Value
+		invoiceKeys := make(map[CircuitKey]struct{})
```

Comment:
I'd keep it this way as the value of `invoiceKeys` is later on passed to an `InvoiceStateAMP` struct which I don't think we should refactor in this PR to avoid scope creep. Noted though for a later change.